14.0 - [REF] account: Adding unittests for concurrency issues in account_move sequences

moylop260 opened the pull request:
Related to https://github.com/odoo/odoo/issues/90465

Reproducing all the new cases of concurrency for Odoo v14.0 account.move records in the following cases:
- Creating draft invoices
- Creating payments
- Deadlock payment vs invoice
- Editing last invoice and creating new one
- Editing last payment and creating new one
- Reconciling last invoice and creating new one
- Reconciling last payment and creating new one
- No be able to configure to standard sequences anymore

All these kind of errors are raising other kind of errors related to subscription in payment exception, sale order in draft but with transaction, pos order errors, ecommerce users raising errors after pay then request the pay again and again and so on

All workers are used waiting for releases in the queue

Issue describing this:
 - https://github.com/odoo/odoo/issues/91873

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/91525).

moylop260 commented:
FYI the builds logs is reproducing concurrency errors and deadlock:
 - https://runbot.odoo.com/runbot/batch/771319/build/15578956

<details>
  <summary>logs</summary>
  
```log

2022-05-17 04:13:00,708 15 INFO 15578990-14-0-all odoo.addons.account.tests.test_sequence_mixin: Starting TestSequenceMixin.test_sequence_concurency_draft_invoices ... 
2022-05-17 04:13:03,093 15 ERROR 15578990-14-0-all odoo.sql_db: bad query: 
            UPDATE account_move SET write_date = write_date WHERE id = (
                SELECT id FROM account_move
                WHERE journal_id = 2 AND name != '/' AND date_trunc('month', date::timestamp without time zone) = date_trunc('month', '2022-05-17'::date)  AND move_type NOT IN ('out_refund', 'in_refund') 
                AND sequence_prefix = (SELECT sequence_prefix FROM account_move WHERE journal_id = 2 AND name != '/' AND date_trunc('month', date::timestamp without time zone) = date_trunc('month', '2022-05-17'::date)  AND move_type NOT IN ('out_refund', 'in_refund')  ORDER BY id DESC LIMIT 1)
                ORDER BY sequence_number DESC
                LIMIT 1
            )
            RETURNING name;
        
ERROR: canceling statement due to statement timeout
CONTEXT:  while updating tuple (33,4) in relation "account_move"
 
2022-05-17 04:13:03,151 15 INFO 15578990-14-0-all odoo.models.unlink: User #1 deleted account.move.line records with IDs: [2286, 2287] 
2022-05-17 04:13:03,161 15 INFO 15578990-14-0-all odoo.models.unlink: User #1 deleted mail.message records with IDs: [3030, 3029] 
2022-05-17 04:13:03,173 15 INFO 15578990-14-0-all odoo.models.unlink: User #1 deleted account.move records with IDs: [727] 
2022-05-17 04:13:03,179 15 INFO 15578990-14-0-all odoo.models.unlink: User #1 deleted mail.followers records with IDs: [3443, 3444] 
2022-05-17 04:13:03,182 15 INFO 15578990-14-0-all odoo.addons.account.tests.test_sequence_mixin: ====================================================================== 
2022-05-17 04:13:03,182 15 ERROR 15578990-14-0-all odoo.addons.account.tests.test_sequence_mixin: FAIL: TestSequenceMixin.test_sequence_concurency_draft_invoices
Traceback (most recent call last):
  File "/data/build/odoo/addons/account/tests/test_sequence_mixin.py", line 515, in test_sequence_concurency_draft_invoices
    invoice2 = self._create_invoice_form(env2, date, post=False)
psycopg2.errors.QueryCanceled: canceling statement due to statement timeout
CONTEXT:  while updating tuple (33,4) in relation "account_move"


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/build/odoo/addons/account/tests/test_sequence_mixin.py", line 518, in test_sequence_concurency_draft_invoices
    self.assertFalse(True, "Should it raises error to user and rollback the whole transaction? %s" % e)
AssertionError: True is not false : Should it raises error to user and rollback the whole transaction? canceling statement due to statement timeout
CONTEXT:  while updating tuple (33,4) in relation "account_move"

 
2022-05-17 04:13:03,186 15 INFO 15578990-14-0-all odoo.addons.account.tests.test_sequence_mixin: Starting TestSequenceMixin.test_sequence_concurency_editing_last_invoice ... 
2022-05-17 04:13:05,356 15 ERROR 15578990-14-0-all odoo.sql_db: bad query: 
            UPDATE account_move SET write_date = write_date WHERE id = (
                SELECT id FROM account_move
                WHERE journal_id = 2 AND name != '/' AND date_trunc('month', date::timestamp without time zone) = date_trunc('month', '2016-01-01'::date)  AND move_type NOT IN ('out_refund', 'in_refund') 
                AND sequence_prefix = (SELECT sequence_prefix FROM account_move WHERE journal_id = 2 AND name != '/' AND date_trunc('month', date::timestamp without time zone) = date_trunc('month', '2016-01-01'::date)  AND move_type NOT IN ('out_refund', 'in_refund')  ORDER BY id DESC LIMIT 1)
                ORDER BY sequence_number DESC
                LIMIT 1
            )
            RETURNING name;
        
ERROR: canceling statement due to statement timeout
CONTEXT:  while updating tuple (34,12) in relation "account_move"
 
2022-05-17 04:13:05,406 15 INFO 15578990-14-0-all odoo.models.unlink: User #1 deleted account.move.line records with IDs: [2290, 2291] 
2022-05-17 04:13:05,413 15 INFO 15578990-14-0-all odoo.models.unlink: User #1 deleted mail.message records with IDs: [3033, 3032] 
2022-05-17 04:13:05,422 15 INFO 15578990-14-0-all odoo.models.unlink: User #1 deleted account.move records with IDs: [729] 
2022-05-17 04:13:05,426 15 INFO 15578990-14-0-all odoo.models.unlink: User #1 deleted mail.followers records with IDs: [3446, 3447] 
2022-05-17 04:13:05,429 15 INFO 15578990-14-0-all odoo.addons.account.tests.test_sequence_mixin: ====================================================================== 
2022-05-17 04:13:05,429 15 ERROR 15578990-14-0-all odoo.addons.account.tests.test_sequence_mixin: FAIL: TestSequenceMixin.test_sequence_concurency_editing_last_invoice
Traceback (most recent call last):
  File "/data/build/odoo/addons/account/tests/test_sequence_mixin.py", line 596, in test_sequence_concurency_editing_last_invoice
    self._create_invoice_form(env1, date)
psycopg2.errors.QueryCanceled: canceling statement due to statement timeout
CONTEXT:  while updating tuple (34,12) in relation "account_move"


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/build/odoo/addons/account/tests/test_sequence_mixin.py", line 598, in test_sequence_concurency_editing_last_invoice
    self.assertFalse(True, "Should it raises error to user and rollback the whole transaction? %s" % e)
AssertionError: True is not false : Should it raises error to user and rollback the whole transaction? canceling statement due to statement timeout
CONTEXT:  while updating tuple (34,12) in relation "account_move"

 
2022-05-17 04:13:05,432 15 INFO 15578990-14-0-all odoo.addons.account.tests.test_sequence_mixin: Starting TestSequenceMixin.test_sequence_concurency_editing_last_payment ... 
2022-05-17 04:13:06,891 15 ERROR 15578990-14-0-all odoo.sql_db: bad query: 
            UPDATE account_move SET write_date = write_date WHERE id = (
                SELECT id FROM account_move
                WHERE journal_id = 9 AND name != '/'
                AND sequence_prefix = (SELECT sequence_prefix FROM account_move WHERE journal_id = 9 AND name != '/' ORDER BY id DESC LIMIT 1)
                ORDER BY sequence_number DESC
                LIMIT 1
            )
            RETURNING name;
        
ERROR: canceling statement due to statement timeout
CONTEXT:  while updating tuple (35,7) in relation "account_move"
 
2022-05-17 04:13:06,899 15 INFO 15578990-14-0-all odoo.addons.account.tests.test_sequence_mixin: ====================================================================== 
2022-05-17 04:13:06,899 15 ERROR 15578990-14-0-all odoo.addons.account.tests.test_sequence_mixin: FAIL: TestSequenceMixin.test_sequence_concurency_editing_last_payment
Traceback (most recent call last):
  File "/data/build/odoo/odoo/api.py", line 793, in get
    return field_cache[record._ids[0]]
KeyError: <NewId 0x7f97d953a130>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/build/odoo/odoo/fields.py", line 972, in __get__
    value = env.cache.get(record, self)
odoo.exceptions.CacheMiss: 'account.payment(<NewId 0x7f97d953a130>,).name'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/build/odoo/addons/account/tests/test_sequence_mixin.py", line 622, in test_sequence_concurency_editing_last_payment
    self._create_payment_form(env1, date)
psycopg2.errors.QueryCanceled: canceling statement due to statement timeout
CONTEXT:  while updating tuple (35,7) in relation "account_move"


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/build/odoo/addons/account/tests/test_sequence_mixin.py", line 624, in test_sequence_concurency_editing_last_payment
    self.assertFalse(True, "Should it raises error to user and rollback the whole transaction? %s" % e)
AssertionError: True is not false : Should it raises error to user and rollback the whole transaction? canceling statement due to statement timeout
CONTEXT:  while updating tuple (35,7) in relation "account_move"

 
2022-05-17 04:13:06,902 15 INFO 15578990-14-0-all odoo.addons.account.tests.test_sequence_mixin: ====================================================================== 
2022-05-17 04:13:06,903 15 ERROR 15578990-14-0-all odoo.addons.account.tests.test_sequence_mixin: ERROR: TestSequenceMixin.test_sequence_concurency_editing_last_payment
Traceback (most recent call last):
  File "/data/build/odoo/odoo/api.py", line 793, in get
    return field_cache[record._ids[0]]
KeyError: 78

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/build/odoo/odoo/fields.py", line 972, in __get__
    value = env.cache.get(record, self)
  File "/data/build/odoo/odoo/api.py", line 796, in get
    raise CacheMiss(record, field)
odoo.exceptions.CacheMiss: 'account.move(78,).edi_show_cancel_button'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/build/odoo/odoo/api.py", line 793, in get
    return field_cache[record._ids[0]]
KeyError: 78

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/build/odoo/odoo/fields.py", line 972, in __get__
    value = env.cache.get(record, self)
  File "/data/build/odoo/odoo/api.py", line 796, in get
    raise CacheMiss(record, field)
odoo.exceptions.CacheMiss: 'account.move(78,).state'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/build/odoo/odoo/fields.py", line 1021, in __get__
    self.compute_value(recs)
  File "/data/build/odoo/odoo/fields.py", line 1177, in compute_value
    records._compute_field_value(self)
  File "/data/build/odoo/addons/mail/models/mail_thread.py", line 410, in _compute_field_value
    return super()._compute_field_value(field)
  File "/data/build/odoo/odoo/models.py", line 4069, in _compute_field_value
    getattr(self, field.compute)()
  File "/data/build/odoo/addons/account_edi/models/account_move.py", line 83, in _compute_edi_show_cancel_button
    if move.state != 'posted':
  File "/data/build/odoo/odoo/fields.py", line 1002, in __get__
    raise MissingError("\n".join([
odoo.exceptions.MissingError: Record does not exist or has been deleted.
(Record: account.move(78,), User: 1)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/build/odoo/odoo/api.py", line 793, in get
    return field_cache[record._ids[0]]
KeyError: 78

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/build/odoo/odoo/fields.py", line 972, in __get__
    value = env.cache.get(record, self)
  File "/data/build/odoo/odoo/api.py", line 796, in get
    raise CacheMiss(record, field)
odoo.exceptions.CacheMiss: 'account.move(78,).state'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/build/odoo/addons/account/tests/test_sequence_mixin.py", line 465, in _clean_moves
    moves.button_draft()
  File "/data/build/odoo/addons/stock_account/models/account_move.py", line 59, in button_draft
    res = super(AccountMove, self).button_draft()
  File "/data/build/odoo/addons/membership/models/account_move.py", line 14, in button_draft
    res = super(AccountMove, self).button_draft()
  File "/data/build/odoo/addons/account_edi/models/account_move.py", line 349, in button_draft
    if move.edi_show_cancel_button:
  File "/data/build/odoo/odoo/fields.py", line 1023, in __get__
    self.compute_value(record)
  File "/data/build/odoo/odoo/fields.py", line 1177, in compute_value
    records._compute_field_value(self)
  File "/data/build/odoo/addons/mail/models/mail_thread.py", line 410, in _compute_field_value
    return super()._compute_field_value(field)
  File "/data/build/odoo/odoo/models.py", line 4069, in _compute_field_value
    getattr(self, field.compute)()
  File "/data/build/odoo/addons/account_edi/models/account_move.py", line 83, in _compute_edi_show_cancel_button
    if move.state != 'posted':
  File "/data/build/odoo/odoo/fields.py", line 1002, in __get__
    raise MissingError("\n".join([
odoo.exceptions.MissingError: Record does not exist or has been deleted.
(Record: account.move(78,), User: 1)
 
2022-05-17 04:13:06,906 15 INFO 15578990-14-0-all odoo.addons.account.tests.test_sequence_mixin: Starting TestSequenceMixin.test_sequence_concurency_pay2inv_inv2pay ... 
2022-05-17 04:13:09,887 15 ERROR 15578990-14-0-all odoo.sql_db: bad query: 
            UPDATE account_move SET write_date = write_date WHERE id = (
                SELECT id FROM account_move
                WHERE journal_id = 2 AND name != '/' AND date_trunc('month', date::timestamp without time zone) = date_trunc('month', '2022-05-17'::date)  AND move_type NOT IN ('out_refund', 'in_refund') 
                AND sequence_prefix = (SELECT sequence_prefix FROM account_move WHERE journal_id = 2 AND name != '/' AND date_trunc('month', date::timestamp without time zone) = date_trunc('month', '2022-05-17'::date)  AND move_type NOT IN ('out_refund', 'in_refund')  ORDER BY id DESC LIMIT 1)
                ORDER BY sequence_number DESC
                LIMIT 1
            )
            RETURNING name;
        
ERROR: deadlock detected
DETAIL:  Process 3216393 waits for ShareLock on transaction 2339227478; blocked by process 3216391.
Process 3216391 waits for ShareLock on transaction 2339227476; blocked by process 3216393.
HINT:  See server log for query details.
CONTEXT:  while updating tuple (34,20) in relation "account_move"
 
Exception in thread Thread payment invoice:
Traceback (most recent call last):
  File "/usr/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/data/build/odoo/addons/account/tests/test_sequence_mixin.py", line 537, in _create_invoice_payment
    raise exc
  File "/data/build/odoo/addons/account/tests/test_sequence_mixin.py", line 529, in _create_invoice_payment
    self._create_invoice_form(env, date)
  File "/data/build/odoo/addons/account/tests/test_sequence_mixin.py", line 437, in _create_invoice_form
    with Form(env['account.move'].with_context(default_move_type='out_invoice')) as invoice_form:
  File "/data/build/odoo/odoo/tests/common.py", line 1650, in __init__
    self._init_from_defaults(self._model)
  File "/data/build/odoo/odoo/tests/common.py", line 1732, in _init_from_defaults
    self._perform_onchange([])
  File "/data/build/odoo/odoo/tests/common.py", line 2018, in _perform_onchange
    result = record.onchange(self._onchange_values(), fields, spec)
  File "/data/build/odoo/addons/account/models/account_move.py", line 1125, in onchange
    return super(AccountMove, self.with_context(recursive_onchanges=False)).onchange(values, field_name, field_onchange)
  File "/data/build/odoo/odoo/models.py", line 6264, in onchange
    record._onchange_eval(name, field_onchange[name], result)
  File "/data/build/odoo/odoo/models.py", line 6006, in _onchange_eval
    method_res = method(self)
  File "/data/build/odoo/addons/account/models/account_move.py", line 429, in _onchange_journal
    if self.state == 'draft' and self._get_last_sequence() and self.name and self.name != '/':
  File "/data/build/odoo/addons/account/models/sequence_mixin.py", line 179, in _get_last_sequence
    self.env.cr.execute(query, param)
  File "<decorator-gen-3>", line 2, in execute
  File "/data/build/odoo/odoo/sql_db.py", line 101, in check
    return f(self, *args, **kwargs)
  File "/data/build/odoo/odoo/sql_db.py", line 298, in execute
    res = self._obj.execute(query, params)
psycopg2.errors.DeadlockDetected: deadlock detected
DETAIL:  Process 3216393 waits for ShareLock on transaction 2339227478; blocked by process 3216391.
Process 3216391 waits for ShareLock on transaction 2339227476; blocked by process 3216393.
HINT:  See server log for query details.
CONTEXT:  while updating tuple (34,20) in relation "account_move"

2022-05-17 04:13:12,121 15 INFO 15578990-14-0-all odoo.models.unlink: User #1 deleted account.move.line records with IDs: [2294, 2295, 2296, 2297] 
2022-05-17 04:13:12,132 15 INFO 15578990-14-0-all odoo.models.unlink: User #1 deleted mail.message records with IDs: [3038, 3037, 3036] 
2022-05-17 04:13:12,165 15 INFO 15578990-14-0-all odoo.models.unlink: User #1 deleted account.move records with IDs: [731, 732] 
2022-05-17 04:13:12,170 15 INFO 15578990-14-0-all odoo.models.unlink: User #1 deleted mail.followers records with IDs: [3450, 3451, 3452] 
2022-05-17 04:13:12,174 15 INFO 15578990-14-0-all odoo.addons.account.tests.test_sequence_mixin: ====================================================================== 
2022-05-17 04:13:12,175 15 ERROR 15578990-14-0-all odoo.addons.account.tests.test_sequence_mixin: FAIL: TestSequenceMixin.test_sequence_concurency_pay2inv_inv2pay
Traceback (most recent call last):
  File "/data/build/odoo/addons/account/tests/test_sequence_mixin.py", line 568, in test_sequence_concurency_pay2inv_inv2pay
    raise self.last_thread_exc
psycopg2.errors.DeadlockDetected: deadlock detected
DETAIL:  Process 3216393 waits for ShareLock on transaction 2339227478; blocked by process 3216391.
Process 3216391 waits for ShareLock on transaction 2339227476; blocked by process 3216393.
HINT:  See server log for query details.
CONTEXT:  while updating tuple (34,20) in relation "account_move"


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/build/odoo/addons/account/tests/test_sequence_mixin.py", line 570, in test_sequence_concurency_pay2inv_inv2pay
    self.assertFalse(True, "Should it raises deadlock error to user and rollback the whole 2 transactions? %s" % e)
AssertionError: True is not false : Should it raises deadlock error to user and rollback the whole 2 transactions? deadlock detected
DETAIL:  Process 3216393 waits for ShareLock on transaction 2339227478; blocked by process 3216391.
Process 3216391 waits for ShareLock on transaction 2339227476; blocked by process 3216393.
HINT:  See server log for query details.
CONTEXT:  while updating tuple (34,20) in relation "account_move"

 
2022-05-17 04:13:12,178 15 INFO 15578990-14-0-all odoo.addons.account.tests.test_sequence_mixin: Starting TestSequenceMixin.test_sequence_concurency_payments ... 
2022-05-17 04:13:13,910 15 ERROR 15578990-14-0-all odoo.sql_db: bad query: 
            UPDATE account_move SET write_date = write_date WHERE id = (
                SELECT id FROM account_move
                WHERE journal_id = 9 AND name != '/'
                AND sequence_prefix = (SELECT sequence_prefix FROM account_move WHERE journal_id = 9 AND name != '/' ORDER BY id DESC LIMIT 1)
                ORDER BY sequence_number DESC
                LIMIT 1
            )
            RETURNING name;
        
ERROR: canceling statement due to statement timeout
CONTEXT:  while updating tuple (36,32) in relation "account_move"
 
2022-05-17 04:13:13,956 15 INFO 15578990-14-0-all odoo.models.unlink: User #1 deleted account.move.line records with IDs: [2304, 2305] 
2022-05-17 04:13:13,963 15 INFO 15578990-14-0-all odoo.models.unlink: User #1 deleted mail.message records with IDs: [3046] 
2022-05-17 04:13:13,995 15 INFO 15578990-14-0-all odoo.models.unlink: User #1 deleted account.move records with IDs: [736] 
2022-05-17 04:13:13,999 15 INFO 15578990-14-0-all odoo.models.unlink: User #1 deleted mail.followers records with IDs: [3460] 
2022-05-17 04:13:14,003 15 INFO 15578990-14-0-all odoo.addons.account.tests.test_sequence_mixin: ====================================================================== 
2022-05-17 04:13:14,003 15 ERROR 15578990-14-0-all odoo.addons.account.tests.test_sequence_mixin: FAIL: TestSequenceMixin.test_sequence_concurency_payments
Traceback (most recent call last):
  File "/data/build/odoo/odoo/api.py", line 793, in get
    return field_cache[record._ids[0]]
KeyError: <NewId 0x7f97d86e14c0>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/build/odoo/odoo/fields.py", line 972, in __get__
    value = env.cache.get(record, self)
odoo.exceptions.CacheMiss: 'account.payment(<NewId 0x7f97d86e14c0>,).name'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/build/odoo/addons/account/tests/test_sequence_mixin.py", line 490, in test_sequence_concurency_payments
    self._create_payment_form(env2, date)
psycopg2.errors.QueryCanceled: canceling statement due to statement timeout
CONTEXT:  while updating tuple (36,32) in relation "account_move"


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/data/build/odoo/addons/account/tests/test_sequence_mixin.py", line 492, in test_sequence_concurency_payments
    self.assertFalse(True, "Should it raises error to user and rollback the whole transaction? %s" % e)
AssertionError: True is not false : Should it raises error to user and rollback the whole transaction? canceling statement due to statement timeout
CONTEXT:  while updating tuple (36,32) in relation "account_move"
```

</details>

luisg123v commented on addons/account/tests/test_sequence_mixin.py:
```diff
@@ -427,3 +432,193 @@ def test_change_journal_on_first_account_move(self):
+            invoice_form.partner_id = env.ref('base.res_partner_12')
+            invoice_form.invoice_date = date
+            with invoice_form.invoice_line_ids.new() as line_form:
+                line_form.name = 'concurency_test_invoice'
```
```suggestion
                line_form.name = 'concurrency_test_invoice'
```

luisg123v commented on addons/account/tests/test_sequence_mixin.py:
```diff
@@ -427,3 +432,193 @@ def test_change_journal_on_first_account_move(self):
+            moves.unlink()
+            env.cr.commit()
+
+    def test_sequence_concurency_payments(self):
```
```suggestion
    def test_sequence_concurrency_payments(self):
```

luisg123v commented on addons/account/tests/test_sequence_mixin.py:
```diff
@@ -427,3 +432,193 @@ def test_change_journal_on_first_account_move(self):
+
+    def test_sequence_concurency_draft_invoices(self):
+        """Creating 2 DRAFT invoices not should raises errors"""
+        date = fields.Date.from_string('2016-01-01')
```
`from_string` is deprecated in favor of `to_date`

luisg123v commented on addons/account/tests/test_sequence_mixin.py:
```diff
@@ -427,3 +432,193 @@ def test_change_journal_on_first_account_move(self):
+            except (psycopg2.errors.QueryCanceled, psycopg2.errors.TransactionRollbackError) as e:
+                self.assertFalse(True, "Should it raises error to user and rollback the whole transaction? %s" % e)
+
+    def test_sequence_concurency_draft_invoices(self):
```
Same typo

luisg123v commented on addons/account/tests/test_sequence_mixin.py:
```diff
@@ -427,3 +432,193 @@ def test_change_journal_on_first_account_move(self):
+        return invoice
+
+    def _create_payment_form(self, env, date):
+        with Form(env['account.payment'].with_context(**{
```
What do you think about first saving the context to a variable and then `with_context(**ctx)`?

luisg123v reviewed (changes requested):
Typo in title: unitests -> unittests

moylop260 commented:
FYI I have created this unittest dual-compatibles with v13.0 and v14.0

You can see the runbot green for v13.0 for all the same cases where v14.0 is red:
 - https://github.com/odoo/odoo/pull/91614

moylop260 commented:
Should I add unittests for Bill Invoices?

I mean, the result will be the same than Customer Invoices but I'm not sure if your legal basis talks about Bill Invoices number no-gap too

moylop260 commented:
cc @william-andre

guewen commented:
I ended up here after we issued the same issue on our database freshly migrated to Odoo 14.0 (from 12.0).
From [Twitter](https://twitter.com/guewenb/status/1539514978042007553) and https://github.com/odoo/odoo/issues/90465, I see we are not the only one impacted, which doesn't surprise me, because I don't see how the new sequences model would be able to scale.

The 4 first days after the migration were insufferable, with a huge amount of transaction rollback errors on all kind of actions.
I say 4 first, days, because on the 5th day, we installed the OCA module [account_move_name_sequence](https://github.com/OCA/account-financial-tools/tree/14.0/account_move_name_sequence) which made things back to normal.

@moylop260 did a very good case at documenting the use cases that provokes concurrent writes, I ask you to have an honest read of the documentation of  [account_move_name_sequence](https://github.com/OCA/account-financial-tools/tree/14.0/account_move_name_sequence).

To give an example, which is actually not in the list of use cases but happens : even when only **showing the form view** of an invoice, we sometimes had "could not serialize" errors. Why? Because the computation of `AccountMove.highest_name` (a non-stored field) calls `record._get_last_sequence()`, method that executes an UPDATE on the last `account_move`. https://github.com/odoo/odoo/blob/1f2af8abe9b1d4781f1758a98cc8a94bd709cf1c/addons/account/models/account_move.py#L1227-L1230

This is only a single issue, but it makes me wonder how far this design was thought out, how it could have landed in production, and most importantly, why you would refuse to reconsider your approach. Sorry if I'm a bit harsh, but that's a fact : it cannot be used on an ERP with some volume. I can't thank enough the OCA, specifically @alexis-via and @moylop260 for the work on `account_move_name_sequence` who saved our day.

## Data to support my words:

Number of rows in  `account_move` : 11'900'000+
Account moves created per day : average of 5000


Logs that contains "could not serialize" ("Deploy" indicates the installation of the OCA module) during the first week:

![coul_not_serialize](https://user-images.githubusercontent.com/417223/176406783-eb02bbf5-92a1-48ea-992e-ba913edd2313.png)

Second week so far, using the OCA module (I didn't include the week-end):

![could_not_serialize_after_oca_module](https://user-images.githubusercontent.com/417223/176407072-62a0b917-51d9-408d-9602-29724bfacd92.png)

The first week, almost all of the results match this query:

https://github.com/odoo/odoo/blob/1f2af8abe9b1d4781f1758a98cc8a94bd709cf1c/addons/account/models/sequence_mixin.py#L164-L171

The much lower number of hits on the second week are on other (maybe expected) use cases.

Another way to look at this, is comparing the number of moves created during a day, and compare to how much of these errors we had during the same day. The 22nd of June, for some reason was a light in terms of moves creation, only 2619.

![image](https://user-images.githubusercontent.com/417223/176411597-a9a6a23e-18e1-4110-95a6-409a32105311.png)

However, we have 9370 hits on the error during the day, which means 3.5x the number of moves created!

![image](https://user-images.githubusercontent.com/417223/176411019-6c9fd1ba-0413-4384-8e1e-3c2f31a2d9c2.png)

Hope this data can encourage you to move forward on this subject.
Thanks for reading.

william-andre commented:
@guewen This comment was helpful, here is something that should improve the behavior already
https://github.com/odoo/odoo/pull/94852

guewen commented:
Thank you @william-andre

moylop260 commented:
Thanks @william-andre 

I just rebased my branch with new fix committed

I saw the first test `TestSequenceConcurrency.test_sequence_concurrency_10_draft_invoices ... `  is now green

Thank you so much for taking care about this

Almost of the other tests are still failing

We understand the stable policies and it is a good fix for stable branch

Is there plan to fix the other cases for future new version v16.0?

william-andre commented:
I'm glad it helped. Future improvements should still be coming; it would be helpful to know if most of the issues are gone with the previous fix. I a large portion of the failed transactions to be because of that.

Concerning the deadlocks, I got the fact that it can happen theoretically, but I have no idea of what functional flow would lead you to create a payment followed by an invoice. At least not for something that happens on a daily (even weekly) basis.
The only way it could work anyways is with a gap sequence, which we cannot change (and I'm pretty sure that you wouldn't want that either even if you think it doesn't matter to you)

So if any of you can provide me with the impact of the last PR on the number of transactions retried/aborted/dealocked, that would be very much appreciated.

moylop260 commented:
> it would be helpful to know if most of the issues are gone with the previous fix. I a large portion of the failed transactions to be because of that.

Sorry, I have already deployed in production 14.0 the OCA module [account_move_name_sequence](https://github.com/OCA/account-financial-tools/tree/14.0/account_move_name_sequence).

So, the concurrency errors are not reproduced anymore in production for our case

But the unittest that I have created in this PR were inspired by production logs

As you could see the draft invoice concurrency issue reported by Guewen it was done in the tests too

The slow reconciliation is not hard to reproduce

Create 1k posted invoices and only 1 payment post and reconcile
It will spent a long time locking the last records of invoice and payment
if you try to create a new payment or invoice it will be locked for a long time too so no one can post a new payment or invoice for a while

They are reproduced in the unittests:
 - test_sequence_concurrency_40_reconciling_last_invoice
 - test_sequence_concurrency_50_reconciling_last_payment


> but I have no idea of what functional flow would lead you to create a payment followed by an invoice

Subscription:
 - https://github.com/odoo/enterprise/blob/d6e0d7283b4534c2e5a4ea73012ef8f22b13e786/sale_subscription/models/sale_subscription.py#L1048-L1060

Sale:
 - https://github.com/odoo/odoo/blob/dee53d8aecd3d1eedd37bfa93c07ca28a3441fbe/addons/sale/models/payment.py#L115-L125


They are good cases for locking invoices and payments

Maybe custom modules could have extra ways to auto-generate invoices or payments without deadlocks for <=v13.0 but >=v14.0

moylop260 commented:
> But when do you create 1k invoices, 1 payment and then reconcile in the same transaction?

Sorry, I was talking about locking the last invoice/payment for a long time (e.g. reconcile process) then a different transaction to confirm new invoices/payments
They will not be able to confirm new invoices/payments since that last record is already locked and the "post" are trying to lock them too

> it won't lock more than before with a no gap sequence.

In fact, If you see the same unittest for v13.0 that I have created for you:
 - https://github.com/odoo/odoo/pull/91614

You will realize that the no-gap sequence is generating green result for the following unittests:
 - test_sequence_concurrency_40_reconciling_last_invoice
 - test_sequence_concurrency_50_reconciling_last_payment

Because it is not locking the last record but locking extra table only posting so no matter if you are reconciling (locking) the last invoice/payment for a long time because no-gap only will lock if you post too many invoices at the same time but not if you only changed something in the last invoice.

moylop260 commented:
I mean, the important piece of code from unittest:
 - test_sequence_concurrency_40_reconciling_last_invoice
 - test_sequence_concurrency_50_reconciling_last_payment

Are the following lines:
 - <img width="444" alt="Screen Shot 2022-07-06 at 21 07 10" src="https://user-images.githubusercontent.com/6644187/177674865-a9c3db4d-407c-42b2-8902-2058a0f32c4a.png">

Notice It is using 2 cursors/transactions

Using the same code of the unittest for v13.0 and v14.0 you can see

Result for v13.0 same unittest:
 - <img width="1206" alt="Screen Shot 2022-07-06 at 21 10 47" src="https://user-images.githubusercontent.com/6644187/177675188-cdd8893d-18f9-453d-9005-0c33123e3215.png">

Result for v14.0 same unittest:
 - <img width="761" alt="Screen Shot 2022-07-06 at 21 14 36" src="https://user-images.githubusercontent.com/6644187/177675577-7e9e8c7e-e443-4ed2-bd4b-8bdbcef24e08.png">

moylop260 commented:
Fixed for >=v16.0 from https://github.com/odoo/odoo/pull/104606
Fixed for >=14.0 and <16.0 from OCA module [account_move_name_sequence](https://github.com/OCA/account-financial-tools/tree/14.0/account_move_name_sequence)
